homeassistant/components/mqtt/__init__.py

```diff
@@ -571,50 +583,35 @@ def _mqtt_on_connect(self, _mqttc, _userdata, _flags, result_code):
             self._mqttc.disconnect()
             return
 
-        self.progress = {}
-        self.subscribed_topics = {}
-        for topic, qos in self.wanted_topics.items():
-            self.hass.add_job(self.async_subscribe, topic, qos)
+        for topic, subs in self._grouped_subscriptions().items():
+            # Re-subscribe with the highest requested qos
+            max_qos = max(subscription.qos for subscription in subs)
+            self.hass.add_job(self._async_perform_subscription, topic, max_qos)
 
         if self.birth_message:
-            self.hass.add_job(self.async_publish(
-                self.birth_message.get(ATTR_TOPIC),
-                self.birth_message.get(ATTR_PAYLOAD),
-                self.birth_message.get(ATTR_QOS),
-                self.birth_message.get(ATTR_RETAIN)))
-
-    def _mqtt_on_subscribe(self, _mqttc, _userdata, mid, granted_qos):
-        """Subscribe successful callback."""
-        topic = self.progress.pop(mid, None)
-        if topic is None:
-            return
-        self.subscribed_topics[topic] = granted_qos[0]
+            self.hass.add_job(self.async_publish(self.birth_message.topic,
+                                                 self.birth_message.payload,
+                                                 self.birth_message.qos,
+                                                 self.birth_message.retain))
 
     def _mqtt_on_message(self, _mqttc, _userdata, msg):
         """Message received callback."""
-        dispatcher_send(
-            self.hass, SIGNAL_MQTT_MESSAGE_RECEIVED, msg.topic, msg.payload,
-            msg.qos
-        )
-
-    def _mqtt_on_unsubscribe(self, _mqttc, _userdata, mid, granted_qos):
-        """Unsubscribe successful callback."""
-        topic = self.progress.pop(mid, None)
-        if topic is None:
-            return
-        self.subscribed_topics.pop(topic, None)
+        _LOGGER.debug("Received message on %s: %s", msg.topic, msg.payload)
```

Comment:
One thing that was nice about the `dispatcher` is that this code would be executed inside the event loop, which is a lot faster when calling `hass.async_add_job` (which happens inside `_run_subscription_callback`). There is quite some overhead to communicate between thread and event loop.

So, you could create a callback to be scheduled to handle matching etc. This would also help with the tests which would call that method directly.

```
    def _mqtt_on_message(self, _mqttc, _userdata, msg):
        self.hass.async_add_job(self._mqtt_handle_message(msg))

    @callback
    def _mqtt_handle_message(self, msg):
        …
```